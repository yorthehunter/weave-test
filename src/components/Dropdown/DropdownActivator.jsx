import React from 'react';
import { string, node, func } from 'prop-types';
import CustomTag from '../utilities/CustomTag';

const componentPrefix = 'weave-dropdown-activator';

const DropdownActivator = (props) => {
  const componentClasses =
    `${componentPrefix}`;

  return (
    <CustomTag
      {...props}
      componentClasses={componentClasses}
      componentStyles={{ cursor: 'pointer' }}
      onClick={props.onClick}
    >
      { props.children }
    </CustomTag>
  );
};

const defaultProps = {
  tag: 'span',
};

const propTypes = {
  tag: string,
  children: node.isRequired,
  onClick: func.isRequired,
};

DropdownActivator.displayName = 'DropdownActivator';
DropdownActivator.defaultProps = defaultProps;
DropdownActivator.propTypes = propTypes;
export default DropdownActivator;
