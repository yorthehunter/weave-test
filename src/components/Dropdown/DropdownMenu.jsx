import React from 'react';
import { string, node, bool } from 'prop-types';
import CustomTag from '../utilities/CustomTag';

const componentPrefix = 'weave-dropdown-menu';

const DropdownMenu = (props) => {
  const {
    children,
    right,
    menuItems,
    dropdownVisible,
  } = props;

  const componentClasses =
    `${componentPrefix}`;

  return dropdownVisible ?
    (
      <CustomTag
        {...props}
        componentClasses={componentClasses}
        componentStyles={right ? { right: 0 } : {}}
      >
        {
          menuItems ?
            React.Children.map(children, child => (
              <li className={`${componentPrefix}__item`}>
                {child}
              </li>
            ))
            :
            children
        }
      </CustomTag>
    )
    :
    null;
};

const defaultProps = {
  tag: 'ul',
  menuItems: true,
  dropdownVisible: false,
  right: false,
};

const propTypes = {
  tag: string,
  menuItems: bool,
  children: node.isRequired,
  dropdownVisible: bool,
  right: bool,
};

DropdownMenu.displayName = 'DropdownMenu';
DropdownMenu.defaultProps = defaultProps;
DropdownMenu.propTypes = propTypes;
export default DropdownMenu;
