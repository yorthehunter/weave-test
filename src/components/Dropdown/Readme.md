Default Dropdown

```js
<Dropdown>
  <DropdownActivator>Click here!</DropdownActivator>
  <DropdownMenu>Dropdown content!</DropdownMenu>
</Dropdown>
```

Dropdown with menu list items

```js
<Dropdown>
  <DropdownActivator>Click here!</DropdownActivator>
  <DropdownMenu>
    <li className="weave-dropdown-menu__item">List item</li>
    <a href="#" className="weave-dropdown-menu__item">List item</a>
    <Separator horizontal />
    <div className="weave-dropdown-menu__item">List item</div>
  </DropdownMenu>
</Dropdown>
```
