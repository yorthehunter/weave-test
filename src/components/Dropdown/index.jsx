import React from 'react';
import { string, bool, node } from 'prop-types';
import CustomTag from '../utilities/CustomTag';

const componentPrefix = 'weave-has-dropdown';

export default class Dropdown extends React.Component {
  constructor() {
    super();
    this.openDropdown = this.openDropdown.bind(this);
    this.handleOutsideClick = this.handleOutsideClick.bind(this);
  }

  componentWillMount() {
    this.setState({ dropdownVisible: false });
  }

  openDropdown(e) {
    e.preventDefault();

    if (!this.state.dropdownVisible) {
      document.addEventListener('click', this.handleOutsideClick, false);
    } else {
      document.removeEventListener('click', this.handleOutsideClick, false);
    }

    this.setState({
      dropdownVisible: !this.state.dropdownVisible,
    });
  }

  handleOutsideClick(e) {
    // Ignore clicks on the component itself.
    if (this.node.contains(e.target)) {
      return;
    }

    this.setState({
      dropdownVisible: false,
    });

    document.removeEventListener('click', this.handleOutsideClick, false);
  }

  renderChildren() {
    return React.Children.map(this.props.children, (child, index) => {
      if (child.type.displayName === 'DropdownMenu' || child.props['data-dropdown']) {
        // Checks children for <Popover> component and passes visibility state
        return React.cloneElement(
          child,
          {
            key: index,
            right: this.props.right,
            dropdownVisible: this.state.dropdownVisible,
          },
        );
      } else if (child.type.displayName === 'DropdownActivator' || child.props['data-dropdown-activator']) {
        // Checks children of a type <Activator> or that have a data attribute of
        // 'data-dropdown-activator' and adds an onClick method that will open the dropdown
        return React.cloneElement(child, { onClick: this.openDropdown });
      }
      // Will return any other children as normal
      return child;
    });
  }

  render() {
    const componentClasses = `${componentPrefix}`;

    return (
      <CustomTag
        {...this.props}
        componentRef={_node => { this.node = _node; }}
        componentClasses={componentClasses}
        componentStyles={this.props.right ? { position: 'relative' } : {}}
      >
        {this.renderChildren()}
      </CustomTag>
    );
  }
}

const defaultProps = {
  tag: 'span',
  right: false,
  children: null,
};

const propTypes = {
  tag: string,
  right: bool,
  children: node,
};

Dropdown.defaultProps = defaultProps;
Dropdown.propTypes = propTypes;
