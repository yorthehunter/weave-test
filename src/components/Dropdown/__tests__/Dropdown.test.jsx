import React from 'react';
import renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Dropdown from '../';
import DropdownActivator from '../DropdownActivator';
import DropdownMenu from '../DropdownMenu';

describe('Dropdown', () => {
  const onClickFunction = jest.fn();

  const component = (
    <Dropdown>
      <DropdownActivator onClick={onClickFunction}>Click here to open dropdown!</DropdownActivator>
      <DropdownMenu>
        <li>Dropdown</li>
        <li>Menu</li>
      </DropdownMenu>
    </Dropdown>
  );
  const wrapper = mount(component);

  it('renders correctly with default values', () => {
    const rendered = renderer.create(component).toJSON();
    expect(rendered).toMatchSnapshot();
  });

  it('opens a DropdownMenu when DropdownActivator is clicked', () => {
    wrapper.find('.weave-dropdown-activator').simulate('click');
    expect(wrapper.find('.weave-dropdown-menu')).toExist();
  });

  it('renders with a right-aligned menu', () => {
    wrapper.setProps({ right: true });
    expect(wrapper.find('.weave-dropdown-menu')).toHaveStyle('right', 0);
  });

  it('should not call action on click inside the component', () => {
    wrapper.find('.weave-dropdown-menu').simulate('click');
    expect(onClickFunction).not.toHaveBeenCalled();
  });

  it('closes a DropdownMenu when DropdownActivator is clicked a second time', () => {
    wrapper.find('.weave-dropdown-activator').simulate('click');
    expect(wrapper.find('.weave-dropdown-menu')).not.toExist();
  });
});

describe('Dropdown', () => {
  it('renders no dropdown if no children are found', () => {
    const wrapper = shallow(<Dropdown />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders children as normal if they are not of type DropdownMenu or DropdownActivator', () => {
    const wrapper = shallow(
      <Dropdown>
        <div>Hello!</div>
      </Dropdown>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
