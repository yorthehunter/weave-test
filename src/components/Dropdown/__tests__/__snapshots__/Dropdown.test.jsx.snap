// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders children as normal if they are not of type DropdownMenu or DropdownActivator 1`] = `
<CustomTag
  componentClasses="weave-has-dropdown"
  componentRef={[Function]}
  componentStyles={Object {}}
  customClasses={null}
  customStyles={Object {}}
  href={null}
  onClick={null}
  right={false}
  tag="span"
  type={null}
  value={null}
>
  <div
    key=".0"
  >
    Hello!
  </div>
</CustomTag>
`;

exports[`Dropdown renders correctly with default values 1`] = `
<span
  className="weave-has-dropdown"
  href={null}
  onClick={null}
  style={Object {}}
  type={null}
  value={null}
>
  <span
    className="weave-dropdown-activator"
    href={null}
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
    type={null}
    value={null}
  >
    Click here to open dropdown!
  </span>
</span>
`;

exports[`Dropdown renders no dropdown if no children are found 1`] = `
<CustomTag
  componentClasses="weave-has-dropdown"
  componentRef={[Function]}
  componentStyles={Object {}}
  customClasses={null}
  customStyles={Object {}}
  href={null}
  onClick={null}
  right={false}
  tag="span"
  type={null}
  value={null}
/>
`;
